ast/compile_test.go

```diff
@@ -4252,3 +4265,279 @@ func TestCompilerPassesTypeCheckNegative(t *testing.T) {
 		t.Fatal("Incorrectly detected a type-checking violation")
 	}
 }
+
+func testParseSchema(t *testing.T, schema, expectedType string) {
+	var sch interface{}
+	err := util.Unmarshal([]byte(schema), &sch)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	newtype, err := setTypesWithSchema(sch)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	if newtype == nil {
+		t.Fatalf("parseSchema returned nil type")
+	}
+	if newtype.String() != expectedType {
+		t.Fatalf("parseSchema returned an incorrect type: %s", newtype.String())
+	}
+}
+
+func TestParseSchemaObject(t *testing.T) {
+	testParseSchema(t, objectSchema, "object<b: array<object<a: number, b: array<number>, c: any>>, foo: string>")
+}
+
+func TestSetTypesWithSchemaRef(t *testing.T) {
+	var sch interface{}
+	err := util.Unmarshal([]byte(podSchema), &sch)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	newtype, err := setTypesWithSchema(sch)
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+	if newtype == nil {
+		t.Fatalf("parseSchema returned nil type")
+	}
+	if newtype.String() == "object<apiVersion: string, kind: string, metadata: any, spec: any, status: any>" {
```

Comment:
Can we assert the expected type instead?

Relatedly, is there a way to have a structured comparison instead of comparing strings?

Reply:
Fixed in commit https://github.com/open-policy-agent/opa/pull/3060/commits/bd984a0fcc39b763449b8ea3616d419eeb925ab5

@srenatus  I changed the comparisons to use `Compare` on types instead of string comparisons. 

The test above is for a schema that contains many references. So when all of those are integrated and parsed, the resulting Rego type is ~100 lines long (when printed on my terminal), so it's not very convenient to have a precise assertion. This test is asserting that we don't get any errors when we parse such a schema, and that the newtype is not the one we would have obtained if we had ignored the refs.

I included another tests for Schema refs that is much shorter, and in that case we have a string comparison for convenience to the expected type.

Pl let me know what you think.

Reply:
Thanks, this is fine 👍 

The crucial, to me, bits are these:
> the newtype is not the one we would have obtained if we had ignored the refs.

from reading the code, I didn't understand the meaning of that string it wasn't supposed to be.